back/engines/commercial/content_builder/spec/acceptance/layouts_spec.rb

```diff
@@ -0,0 +1,406 @@
+
+  context 'when authorized' do
+    let(:user) { create(:admin) }
+
+    before do
+      token = Knock::AuthToken.new(payload: user.to_token_payload).token
+      header 'Authorization', "Bearer #{token}"
+    end
+
+    describe 'GET' do
+      context 'when the layout exists' do
+        let(:layout) { create(:layout) }
+        let(:project_id) { layout.content_buildable_id }
+        let(:code) { layout.code }
+
+        get 'web_api/v1/projects/:project_id/content_builder_layouts/:code' do
+          example_request 'Get one layout by project_id and code' do
+            expect(status).to eq 200
+          end
+        end
+      end
+
+      context 'when the project does not exist' do
+        let(:layout) { create(:layout) }
+        let(:project_id) { 'unknown' }
+        let(:code) { layout.code }
+
+        get 'web_api/v1/projects/:project_id/content_builder_layouts/:code' do
+          example_request 'Try to get a layout of a project when the project does not exist' do
+            expect(status).to eq 404
+          end
+        end
+      end
+
+      context 'when no layout with the given code exists for the given project' do
+        let(:layout) { create(:layout) }
+        let(:project_id) { layout.content_buildable_id }
+        let(:code) { 'unknown' }
+
+        get 'web_api/v1/projects/:project_id/content_builder_layouts/:code' do
+          example_request 'Try to get a layout of a project when the code is unknown' do
+            expect(status).to eq 404
+          end
+        end
+      end
+    end
+
+    describe 'POST' do
+      with_options scope: :content_builder_layout do
+        parameter :enabled, 'Indicates that the layout is enabled.'
+        parameter :craftjs_jsonmultiloc, 'The craftjs layout configuration, as a multiloc string.'
+      end
+
+      context 'when the project does not exist' do
+        let(:layout) { create(:layout) }
+        let(:project_id) { 'unknown' }
+        let(:code) { layout.code }
+
+        post 'web_api/v1/projects/:project_id/content_builder_layouts/:code/upsert' do
+          example_request 'Try to upsert a layout for a project that does not exist' do
+            expect(status).to eq 404
+          end
+        end
+      end
+
+      context 'when the layout does not exist' do
+        let(:project) { create(:project) }
+        let(:project_id) { project.id }
+        let(:code) { 'project_description' }
+        let(:enabled) { true }
+        let :craftjs_jsonmultiloc do
+          {
+            'nl-BE': {
+              ROOT: {
+                type: {
+                  resolvedName: 'Container'
+                }
+              }
+            }
+          }
+        end
+
+        post 'web_api/v1/projects/:project_id/content_builder_layouts/:code/upsert' do
+          example_request 'Create a layout for a project' do
+            expect(status).to eq 201
+            json_response = json_parse(response_body)
+            expect(json_response).to include(
```

Comment:
This is really cool! I didn't know this was possible.

Reply:
😃 Yes, RSpec is amazing.